action.yml

```diff
@@ -316,6 +332,100 @@ runs:
+        fi
+        
+        # Add provider versions from .terraform.lock.hcl if it exists
+        if [ -f ".terraform.lock.hcl" ]; then
+          echo "Including provider versions in cache key..."
+          PROVIDER_HASH=$(grep -A 1 "provider " .terraform.lock.hcl | grep "version" | sort | sha256sum | cut -c1-8)
+          HASH_INPUT="${HASH_INPUT}-p${PROVIDER_HASH}"
+        else
+          # Look for provider version constraints in .tf files
+          echo "Looking for provider version constraints in .tf files..."
+          PROVIDER_HASH=$(find . -name "*.tf" -exec grep -h "required_providers" -A 5 {} \; | grep "version" | sort | sha256sum | cut -c1-8)
+          if [ ! -z "$PROVIDER_HASH" ]; then
+            HASH_INPUT="${HASH_INPUT}-p${PROVIDER_HASH}"
+          fi
+        fi
+        
+        # Add OS and architecture for platform-specific caches
+        HASH_INPUT="${HASH_INPUT}-${{ runner.os }}-${{ runner.arch }}"
+        
+        # Create a short hash for cache identification
+        CONTENT_HASH=$(echo -n "$HASH_INPUT" | sha256sum | cut -c1-12)
+        
+        echo "S3_CACHE_HASH=$CONTENT_HASH" >> $GITHUB_ENV
+        echo "Generated cache hash: $CONTENT_HASH"
+        echo "Cache components: $HASH_INPUT"
+
+    - name: restore-s3-cache
+      shell: bash
+      run: |
+        CACHE_HASH="${{ env.S3_CACHE_HASH }}"
+        BUCKET=$(echo "${{ inputs.cache-dependencies-s3-bucket }}" | sed 's|^s3://||')
+        REGION="${{ inputs.cache-dependencies-s3-region }}"
+        
+        if [ ! -d "$TF_PLUGIN_CACHE_DIR" ]; then
+          echo "Creating cache directory: $TF_PLUGIN_CACHE_DIR"
+          mkdir -p "$TF_PLUGIN_CACHE_DIR"
+        fi
+        
+        echo "Looking for existing cache with hash: $CACHE_HASH"
+        
+        # Check if bucket/prefix exists and we have access
+        if ! aws s3 ls "s3://$BUCKET/artifacts/" --region "$REGION" >/dev/null 2>&1; then
+          echo "No existing caches found (bucket/prefix doesn't exist or no access)"
+          echo "Will create new cache after terraform operations"
+          exit 0
+        fi
+        
+        # Look for any directory that contains this hash (could be timestamped or just hash)
+        EXISTING_CACHE=""
+        
+        # Try to find directories ending with our hash
+        CACHE_CANDIDATES=$(aws s3 ls "s3://$BUCKET/artifacts/" --region "$REGION" 2>/dev/null | awk '{print $2}' | sed 's|/||' | grep "$CACHE_HASH" || echo "")
+        
+        if [ ! -z "$CACHE_CANDIDATES" ]; then
+          # Take the first match (they should all have the same content anyway)
+          EXISTING_CACHE=$(echo "$CACHE_CANDIDATES" | head -1)
+          echo "Found existing cache: $EXISTING_CACHE"
+          
+          S3_CACHE_PATH="s3://$BUCKET/artifacts/$EXISTING_CACHE"
+          
+          echo "Restoring cache from: $S3_CACHE_PATH"
+          if aws s3 sync "$S3_CACHE_PATH" "$TF_PLUGIN_CACHE_DIR" --region "$REGION" --quiet 2>/dev/null; then
```

Comment:
I think this is all what we need for restore, an s3 sync from `inputs.cache-dependencies-s3-bucket` to `$TF_PLUGIN_CACHE_DIR`